onnxruntime/core/providers/cpu/text/string_normalizer.cc

```diff
@@ -319,95 +486,154 @@ Status StringNormalizer::Compute(OpKernelContext* ctx) const {
+      status = output_no_filtering(output_shape);
+    } else {
+      // we need to filter
+      InlinedVector<size_t> filtered_strings_indecies;
+      filtered_strings_indecies.reserve(narrow<size_t>(C));
+
+      for (size_t i = 0, lim = narrow<size_t>(C); i < lim; ++i) {
+        const std::string& s = input_span[i];
+        if (stopwords_.count(s) == 0) {
+          filtered_strings_indecies.push_back(i);
         }
-        ++first;
       }
-      status = CopyCaseAction(filtered_strings.cbegin(), filtered_strings.cend(), ctx, locale, converter,
-                              N, filtered_strings.size(), case_change_action_);
-    } else {
-      // Nothing to filter. Copy input to output and change case if needed
-      status = CopyCaseAction(input_data, input_data + C, ctx, locale, converter, N, C, case_change_action_);
+
+      // According to the spec, if all strings are filtered out
+      // the output must have a shape of {1} with a single empty string.
+      const int64_t filtered_count = std::max<int64_t>(1, narrow<int64_t>(filtered_strings_indecies.size()));
+      output_shape.push_back(filtered_count);
+      status = output_filtered(output_shape, filtered_strings_indecies);
     }
   } else {
-    if (!wstopwords_.empty()) {
-      // Filter input. When no case action is required
-      // we simply store original string references.
-      // Otherwise, we store converted strings.
-      InlinedVector<StrRef> filtered_orignal_strings;
-      InlinedVector<std::string> filtered_cased_strings;
-      filtered_orignal_strings.reserve(C);
-      filtered_cased_strings.reserve(C);
-      auto first = input_data;
-      auto const last = input_data + C;
-      while (first != last) {
-        const std::string& s = *first;
-        std::wstring wstr = converter.from_bytes(s);
-        if (wstr == wconv_error) {
-          // Please do not include the input text in the error message as it could
-          // be deemed as a compliance violation by teams using this operator
-          return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
-                        "Input contains invalid utf8 chars");
-        }
-        locale.ChangeCase(compare_caseaction_, wstr);
-        if (0 == wstopwords_.count(wstr)) {
-          if (case_change_action_ == NONE) {
-            filtered_orignal_strings.push_back(std::cref(s));
-          } else {
-            filtered_cased_strings.push_back(converter.to_bytes(wstr));
-          }
+    if (wstopwords_.empty()) {
+      assert(case_change_action_ != NONE);
+      output_shape.push_back(C);
+      status = output_no_filtering(output_shape);
+    } else {
+      // Case insensitive filtering is performed by converting the input strings
+      // to compare_caseaction_. For that we convert to wchar_t UNICODE.
+      // Otherwise, we need to pull ICU library on all platforms.
+      InlinedVector<size_t> filtered_strings_indecies;
+      filtered_strings_indecies.reserve(narrow<size_t>(C));
```

Comment:
should we just make `C` size_t so we don't need to keep calling `narrow<size_t>(C)`?